web/packages/teleterm/src/mainProcess/profileWatcher/profileWatcher.ts

```diff
@@ -39,25 +41,80 @@ interface ClusterStore {
 /**
  * Watches the specified `tshDirectory` for profile changes.
  * File system events are debounced with a 200 ms delay.
+ *
+ * When the watched directory is removed, the watcher emits a profile change
+ * and enters polling mode (with 1 second interval) until the directory reappears.
  */
 export async function* watchProfiles(
   tshDirectory: string,
   tshClient: TshClient,
   clusterStore: ClusterStore,
   options?: { signal?: AbortSignal }
 ): AsyncGenerator<ProfileChangeSet, void, void> {
-  // eslint-disable-next-line unused-imports/no-unused-vars
-  for await (const _ of debounceWatch(tshDirectory, 200, options?.signal)) {
-    const clusters = await tshClient.listRootClusters();
-    const newClusters = new Map(clusters.map(c => [c.uri, c]));
-    const oldClusters = new Map(
-      clusterStore.getRootClusters().map(c => [c.uri, c])
-    );
-
-    const changes = detectChanges(oldClusters, newClusters);
-    if (changes.length > 0) {
-      yield changes;
+  while (!options?.signal?.aborted) {
+    try {
+      // eslint-disable-next-line unused-imports/no-unused-vars
+      for await (const _ of debounceWatch(tshDirectory, 200, options?.signal)) {
+        const clusters = await tshClient.listRootClusters();
+        const newClusters = new Map(clusters.map(c => [c.uri, c]));
+        const oldClusters = new Map(
+          clusterStore.getRootClusters().map(c => [c.uri, c])
+        );
+
+        const changes = detectChanges(oldClusters, newClusters);
+        if (changes.length > 0) {
+          yield changes;
+        }
+      }
+    } catch (error) {
+      // Check if the error is caused by the tshDirectory being removed.
+      // Detecting directory removal via fs.watch is unreliable:
+      //   - On macOS/Linux, it emits a 'rename' event.
+      //   - On Windows, it may throw an EPERM error.
+      // To reliably detect removal on macOS/Linux, we expect tshClient.listRootClusters()
+      // to fail with a filesystem-related error, allowing us to catch all relevant cases here.
```

Comment:
Are we able to inspect those two kinds of errors here instead of calling `pathExists` on every kind of error that might be thrown within the `try` block?